src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+impl<C> ConsumerStateInner<C> {
+    pub fn complete(self, _deadline: OptionDeadline) -> (OptionDeadline, ConsumerState) {
+        (
+            None.into(),
+            ConsumerState::Complete(ConsumerStateInner {
+                config: self.config,
+                decoder: self.decoder,
+                out: self.out,
+                log_namespace: self.log_namespace,
+                consumer_state: Complete,
+            }),
+        )
+    }
+}
+
+impl ConsumerStateInner<Consuming> {
+    const fn new(
+        config: KafkaSourceConfig,
+        decoder: Decoder,
+        out: SourceSender,
+        log_namespace: LogNamespace,
+    ) -> Self {
+        Self {
+            config,
+            decoder,
+            out,
+            log_namespace,
+            consumer_state: Consuming,
+        }
+    }
+
+    /// Spawn a task on the provided JoinSet to consume the kafka StreamPartitionQueue, and handle acknowledgements for the messages consumed
+    /// Returns a channel sender that can be used to signal that the consumer should stop and drain pending acknowledgements,
+    /// and an AbortHandle that can be used to forcefully end the task.
+    fn consume_partition(
+        &self,
+        join_set: &mut JoinSet<(TopicPartition, PartitionConsumerStatus)>,
+        tp: TopicPartition,
+        consumer: Arc<StreamConsumer<KafkaSourceContext>>,
+        p: StreamPartitionQueue<KafkaSourceContext>,
+        acknowledgements: bool,
+        exit_eof: bool,
+    ) -> (oneshot::Sender<()>, tokio::task::AbortHandle) {
+        // TODO there's probably a better way to pass the keys() config around than cloning the entire config object? But they are tied to the lifetime of the config object, so..this avoids lifetime messes for now
+        let conf = self.config.clone();
+        let decoder = self.decoder.clone();
+        let log_namespace = self.log_namespace;
+        let mut out = self.out.clone();
+
+        let (end_tx, mut end_signal) = oneshot::channel::<()>();
+
+        let handle = join_set.spawn(async move {
+            let mut messages = p.stream();
+            let (finalizer, mut ack_stream) = OrderedFinalizer::<FinalizerEntry>::new(None);
+
+            // finalizer is the entry point for new pending acknowledgements;
+            // when it is dropped, no new messages will be consumed, and the
+            // task will end when it reaches the end of ack_stream
+            let mut finalizer = Some(finalizer);
+
+            let mut status = PartitionConsumerStatus::NormalExit;
+
+            loop {
+                tokio::select!(
+                    _ = &mut end_signal, if finalizer.is_some() => {
+                        finalizer.take();
+                    },
```

Comment:
I think this can avoid the `is_some`, since `take` on a `None` is a no-op:
```suggestion
                    _ = &mut end_signal => {
                        finalizer.take();
                    },
```